### ISSUE TITLE
How to make `get` fast by default?

### ISSUE THREAD

**Issue by mih** [datalad/datalad#2264]
#### What is the problem?

#2261  has an example on `get` bein very slow when obtaining many hundred small files.

I could speed things up considerable by passing `-J 10` or `-J 20`. However, `-J 100` was as slow as a single job. With `-J 20` and my CPUs going full tilt I get a combined total of 30KB/s (not a typo) from smaug.

Is there any sensible default, other than a single job, that would make things faster in general? If there is, why isn't it the general default. If there isn't ... well...

**Comment by mih** [comment]
Maybe the issue lies deeper than that:

```
% datalad -l debug get -J 20 .datalad/metadata/objects
[DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.get.Get'> 
[DEBUG  ] Building doc for <class 'datalad.plugin.Plugin'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
[DEBUG  ] Building doc for <class 'datalad.distribution.subdatasets.Subdatasets'> 
[DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
[DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/tmp/hotnewshit> 
[DEBUG  ] Determine what files match the query to work with 
[INFO   ] Actually getting 8 files 
^CTotal (5 ok out of 8) 97%|██████████████████████████████████████████████████████████████▎ | 25.9K/26.6K [00:20<00:00, 823B/ERROR:                                                                                                                       
Interrupted by user while doing magic: KeyboardInterrupt() [cmd.py:_process_one_line:337] 1.92K/1.92K [00:00<00:00, 5.99MB/s]
```

It doesn't actually return, whereas a single job `get` works fine.

**Comment by yarikoptic** [comment]
I would say 1x or 2x of # of CPUs might be a good default
unfortunately in general we cannot yet just make it a default due to (damn, couldn't locate an old issue, too many open) us (me?) not anyhow handling that situation correctly if files should come from datalad-archives -- then there should be only a single `datalad-archives` per repo and/or some locking mechanism to prevent multiple instances of datalad-archive compete, e.g.:
```shell
$> git annex get -J10 --from=datalad-archives .
get sub-01/func/sub-01_task-balloonanalogrisktask_run-01_events.tsv (from datalad-archives...) 
  To obtain some keys we need to fetch an archive of size 2.4 GB

  PROGRESS-JSON: {"command":"get","wanted":[{"here":false,"uuid":"00000000-0000-0000-0000-000000000001","description":"web"},{"here":false,"uuid":"c8bd3d05-33d4-4b59-9d53-ca7efbdcdd13","description":"yoh@smaug:/mnt/btrfs/datasets/datalad/crawl/openfmri/ds000001"},{"here":false,"uuid":"ce96a357-a40c-4107-9039-77124f428aae","description":"datalad"}],"note":"Unable to access these remotes: web\nTry making some of these repositories available:\n\t00000000-0000-0000-0000-000000000001 -- web\n \tc8bd3d05-33d4-4b59-9d53-ca7efbdcdd13 -- yoh@smaug:/mnt/btrfs/datasets/datalad/crawl/openfmri/ds000001\n \tce96a357-a40c-4107-9039-77124f428aae -- datalad\n","skipped":[],"success":false,"key":"MD5E-s2416078169--53fbcb667531f96e2532723e74ed218d.zip","file":null}
failed
get sub-01/func/sub-01_task-balloonanalogrisktask_run-01_bold.nii.gz (from datalad-archives...) failed
get sub-01/func/sub-01_task-balloonanalogrisktask_run-03_events.tsv (from datalad-archives...) failed
get sub-02/func/sub-02_task-balloonanalogrisktask_run-01_bold.nii.gz (from datalad-archives...) failed
get sub-01/anat/sub-01_inplaneT2.nii.gz (from datalad-archives...) 
  To obtain some keys we need to fetch an archive of size 2.4 GB

  PROGRESS-JSON: {"command":"get","wanted":[{"here":false,"uuid":"00000000-0000-0000-0000-000000000001","description":"web"},{"here":false,"uuid":"c8bd3d05-33d4-4b59-9d53-ca7efbdcdd13","description":"yoh@smaug:/mnt/btrfs/datasets/datalad/crawl/openfmri/ds000001"},{"here":false,"uuid":"ce96a357-a40c-4107-9039-77124f428aae","description":"datalad"}],"note":"Unable to access these remotes: web\nTry making some of these repositories available:\n\t00000000-0000-0000-0000-000000000001 -- web\n \tc8bd3d05-33d4-4b59-9d53-ca7efbdcdd13 -- yoh@smaug:/mnt/btrfs/datasets/datalad/crawl/openfmri/ds000001\n \tce96a357-a40c-4107-9039-77124f428aae -- datalad\n","skipped":[],"success":false,"key":"MD5E-s2416581890--662e0713d0ce42bcdbadb8251b893b8a.tgz","file":null}
failed
get sub-02/func/sub-02_task-balloonanalogrisktask_run-02_events.tsv (from datalad-archives...) failed
get sub-02/func/sub-02_task-balloonanalogrisktask_run-01_events.tsv (from datalad-archives...) failed
get sub-02/func/sub-02_task-balloonanalogrisktask_run-03_bold.nii.gz (from datalad-archives...) failed
get sub-02/func/sub-02_task-balloonanalogrisktask_run-03_events.tsv (from datalad-archives...) failed
get sub-03/anat/sub-03_inplaneT2.nii.gz (from datalad-archives...) failed
get sub-03/func/sub-03_task-balloonanalogrisktask_run-01_events.tsv (from datalad-archives...) failed
...
```
ideas/fixes are welcome.

But in case of the metadata fetching I think it should be safe to assume/provide some `-J auto` which would be the one allocating e.g. 2x number of cores

**Comment by mih** [comment]
One more thing, just to get your feedback. I defined a `demo.datalad.org` as an alias to 127.0.0.1. And then I cloned the repo over SSH and ran a manual `git annex get`.

```
% time git annex get .datalad/metadata
(merging origin/git-annex into git-annex...)
(recording state in git...)
....
git-annex: get: 2 failed
git annex get .datalad/metadata  11.36s user 4.23s system 45% cpu 34.379 total
```

This is virtually the same performance  as a `datalad get`, so we do not add any meaningful overhead here.

For comparison, here is the performance of rsync:

```
% time rsync -caLv demo.datalad.org:/tmp/hotnewshit/.datalad/metadata thaddtsdds
receiving incremental file list
created directory thaddtsdds
...
sent 19,935 bytes  received 12,376,581 bytes  24,793,032.00 bytes/sec
total size is 12,263,969  speedup is 0.99
rsync -caLv demo.datalad.org:/tmp/hotnewshit/.datalad/metadata thaddtsdds  0.10s user 0.08s system 47% cpu 0.382 total
```

That is 100x faster -- not sure if it is cheating and going past the SSH network connection somehow... 100 times...

**Comment by yarikoptic** [comment]
My guess is that rsync just generates a single stream through ssh (imagine tar/untar) whenever git-annex deals with each file separately (even though possibly reusing the connection, so no connection reestablishing for ssh, but only that) e.g. separate wget calls if via http, and those separate rsync runs per each file for ssh:
```

[2018-03-05 11:53:13.056723662] transfer start
[2018-03-05 11:53:13.057399505] call: rsync ["--server","-t","--inplace","-e.Lsf",".","--sender","../../mnt/btrfs/datasets-meta6-4/datalad/crawl/.git/annex/objects/8m/gP/SHA256E-s49112--e89727038bf8f58316b393be2fb4e1a0ba8bb5788026f5c0aabee3746da3af6d.xz/SHA256E-s49112--e89727038bf8f58316b393be2fb4e1a0ba8bb5788026f5c0aabee3746da3af6d.xz"]
SHA256E-s49112--e89727038bf8f58316b393be2fb4e1a0ba8bb5788026f5c0aabee3746da3af6d.xz
         49,112 100%   46.84MB/s    0:00:00 (xfr#1, to-chk=0/1)
[2018-03-05 11:53:13.075758781] feed: ssh ["smaug.datalad.org","-S","../../.git/annex/ssh/smaug.datalad.org","-o","ControlMaster=auto","-o","ControlPersist=yes","-T","git-annex-shell 'transferinfo' '/mnt/btrfs/datasets-meta6-4/datalad/crawl' '--debug' 'SHA256E-s49112--e89727038bf8f58316b393be2fb4e1a0ba8bb5788026f5c0aabee3746da3af6d.xz' --uuid 2da4a6d4-d547-4cb7-9f78-0c12f9ce07a2 '--' 'remoteuuid=54404405-4e10-41da-bc66-b5d961ae92b6' 'associatedfile=objects/08/cn-79e0f8d14282f07a3f7b9e45ff0189.xz' '--'"]
[2018-03-05 11:53:13.075668008] process done ExitSuccess
[2018-03-05 11:53:13.076153032] transfer done
[2018-03-05 11:53:13.081302283] process done ExitSuccess
(checksum...) ok
get objects/08/cn-c8723d8e6a151dd8e3db1fe418dd15.xz (from origin...) 
[2018-03-05 11:53:13.087495824] read: rsync ["--progress","--inplace","--perms","-e","'ssh' 'smaug.datalad.org' '-S' '../../.git/annex/ssh/smaug.datalad.org' '-o' 'ControlMaster=auto' '-o' 'ControlPersist=yes' '-T' 'git-annex-shell ''sendkey'' ''/mnt/btrfs/datasets-meta6-4/datalad/crawl'' ''--debug'' ''SHA256E-s100--6d3302fd8190946bb8aa68f4ebe77dd4549d87b61103a65d1dbd5f6b7882aeda.xz'' --uuid 2da4a6d4-d547-4cb7-9f78-0c12f9ce07a2 ''--'' ''remoteuuid=54404405-4e10-41da-bc66-b5d961ae92b6'' ''unlocked='' ''direct='' ''associatedfile=objects/08/cn-c8723d8e6a151dd8e3db1fe418dd15.xz'' ''--'''","--","dummy:","../../.git/annex/tmp/SHA256E-s100--6d3302fd8190946bb8aa68f4ebe77dd4549d87b61103a65d1dbd5f6b7882aeda.xz"]
[2018-03-05 11:53:13.126342185] process done ExitSuccess
```
The whole architecture of git-annex is built on that (e.g. requests are 1 per file for special remotes etc)... not sure if anything could be done (and I think I have asked but I cannot find it, so I might have not) to batch multiple transfers into a single call, @joeyh ?

**Comment by joeyh** [comment]
Yaroslav Halchenko wrote:
> My guess is that rsync just generates a single stream through ssh (imagine tar/
> untar) whenever git-annex deals with each file separately (even though possibly
> reusing the connection, so no connection reestablishing for ssh, but only that)
> e.g. separate wget calls if via http, and those separate rsync runs per each
> file for ssh:

It depends on the remote, for example S3 is able to keep http
connections open and reuse them. But for something like rsync, where the
whole set of files to be transferred in that TCP connection needs to be
known when starting the rsync process, git-annex's architecture does not
make it practical to bundle together multiple files into a single rsync
call. 

Previously discussed here:
https://git-annex.branchable.com/todo/speed_up_transfers_over_ssh+rsync_--_directly_reuse_the_same_connection__63__/

In general I'd be suspicious of using git-annex with tons of small
files, that is not the use case it's designed for.

-- 
see shy jo

**Comment by joeyh** [comment]
https://git-annex.branchable.com/todo/accellerate_ssh_remotes_with_git-annex-shell_mass_protocol/

Adding that would probably improve speed fairly significantly.
But, it will probably take a solid week's work to get there.

-- 
see shy jo

**Comment by yarikoptic** [comment]
meanwhile in #2268 I did think to set -J to auto where auto is `max(min(3, # of CPUs), 8)` IIRC, but I was not brave enough ;-)  Since this is just a minor improvement over what could be achieved with mass protocol, keeping this open anyways

**Comment by yarikoptic** [comment]
I guess that is just serially without any `-J`?  (reminded me about us talking about smth like #2319)

**Comment by mih** [comment]
Yes, that was serial. Here is parallel (with yesterday's annex):

ssh: `datalad get -J 2 .datalad/metadata  47.47s user 17.13s system 55% cpu 1:56.02 total`
http: `datalad get -J 2 .datalad/metadata  25.88s user 11.04s system 69% cpu 52.888 total`

ssh: `datalad get -J 4 .datalad/metadata  54.47s user 15.67s system 108% cpu 1:04.90 total`
http: `datalad get -J 4 .datalad/metadata  27.15s user 9.26s system 125% cpu 28.952 total`

ssh: `datalad get -J 5 .datalad/metadata  86.73s user 16.71s system 143% cpu 1:11.97 total`
http: `datalad get -J 5 .datalad/metadata  45.61s user 8.06s system 230% cpu 23.247 total`

**Comment by mih** [comment]
After @joeyh pointed out that I would no see much benefits for SSH unless also the server end is running a recent version I did the benchmarking again. This time over a hotel wifi (north america - europe), so the overall speed is much slower. But

`ssh: datalad get .datalad/metadata  10.20s user 4.28s system 8% cpu 2:42.51 total`
`http: datalad get .datalad/metadata  21.49s user 10.15s system 9% cpu 5:30.85 total`

And no, I did not accidentally swap the lines ;-) This was serial processing. Here is parallel:

`ssh: datalad get -J 4 .datalad/metadata  22.19s user 6.38s system 57% cpu 49.709 total`
`http: datalad get -J 4 .datalad/metadata  29.16s user 11.32s system 50% cpu 1:20.68 total`

How to make `get` fast by default? Like this! Thanks @joeyh
